[don't merge] Implement mutate mechanic

ninthworld opened the pull request:
https://user-images.githubusercontent.com/1323803/177071507-49eb9a32-512e-4cc6-a3f0-f5af93ac5ed4.mp4

## Mutate Ability
Mutate is an ability keyword which was first add to MTG in 2020 in the set Ikoria: Lair of Behemoths. Due to the complexity of implementing this mechanic, it has been missing in XMage even since. This PR is the most complete implementation of the mutate mechanic so far, with all common interactions (and most edge-case interactions) tested and working.

### Drawbacks
As stated above, this is a complex mechanic to implement properly, and truthfully, XMage was not originally designed with this type of thing in mind. The majority of cards will work just fine with mutate, and there are several abilities and effects that I have gone through and manually updated to work with mutate. However, it is very likely that there are still cards/abilities/effects out there that implement some custom zone change code that will not inherently work with mutate. And going forward, it is possible that new cards/abilities/effects may not take mutate into account when implementing custom zone change code. Unfortunately, there is no easy fix for this, and a proper refactoring of the code-base to fix this may be too much work for such a limited-use ability.

### Test Coverage:
<details><summary>Expand</summary>

- Cast creature without using mutate ability
- Targeted creature is killed before mutate spell resolves
- Mutate over existing creature to retain the creature's P/T, counters, tapped state, and gain all abilities
- Mutated creature is killed
- Mutated creature is bounced
- Mutated creature is sent to its owners library
- Mutated creature with a God-Eternal over or under dies
- Mutated creature is exiled with Oblivion Ring, then Oblivion Ring is destroyed
- Cast mutate with Volo, Guide to Monsters on the battlefield
- Mutate over creature with transform
- Mutate over creature with imprint
- Mutated creature is phased
- Mutate creature under a morphed creature
- Mutate with number of times mutated
- Mutate with Soulbond, Monstrous, Renowned, Attachments, Regenerate, Persist, Undying, Flip-cards
- Mutate with Mimic Vat, Etrata, Nightmare Shepherd abilities
- Mutate with commander, commander damage, Leadership Vacuum
</details>

Related issues fixed: #6390, #7623

ninthworld committed:
Implement mutate mechanic

ninthworld commented:
I'm really not sure how to implement Brokkos without allowing it to be cast like normal from the graveyard (since MutateAbility is implemented as an alternate casting cost, but you cant alter the cost of an ability that cant be cast)

Alex-Vasile commented:
> I'm really not sure how to implement Brokkos without allowing it to be cast like normal from the graveyard (since MutateAbility is implemented as an alternate casting cost, but you cant alter the cost of an ability that cant be cast)

Could you extend MutateAbility inside Brookos, overriding isAvailable to also include Zone.GRAVEYARD?

Additionally, Flashback also is an static ability which provided an alternate cost for casting specifically from the graveyard, maybe something from there may help.

theelk801 commented:
you probably don't need to extend the mutate ability, just add an arg to the mutate ability itself

Alex-Vasile commented:
Are you supposed to be able to cast for the mutate cost using effects like "exile, and you may play until X"? E.g. [Prosper, Tome-Bound]

theelk801 commented:
yeah, the only situations where you can't use it is when it specifies "without paying its mana cost" or "rather than paying its mana cost"

Alex-Vasile commented:
> yeah, the only situations where you can't use it is when it specifies "without paying its mana cost" or "rather than paying its mana cost"

isAvailable is going to have to be  changed then. It can't be checking zones.

awjackson commented on Mage/src/main/java/mage/abilities/keyword/MutateAbility.java:
```diff
@@ -1,39 +1,191 @@
+                            "Cast " + spell.getLogName() + " with mutate for " + getCosts().getText() + " ?", ability, game)) {
+                        mutateCosts.clearPaid();
+                        TargetCreaturePermanent target = new TargetCreaturePermanent(filter);
+                        target.setNotTarget(true);
```
`setNotTarget` is wrong. Creature spells cast for their mutate cost are targetted spells in every sense, just like Auras. For example, you can't mutate onto a creature with shroud or with protection from one of the mutate card's characteristics.

ninthworld replied:
Woops, good catch!

ninthworld replied:
Fixed by reworking `MutateAbility` to extend `SpellAbility` for all of the casting logic.

awjackson commented on Mage/src/main/java/mage/abilities/keyword/MutateAbility.java:
```diff
@@ -1,39 +1,191 @@
+                        if (target.canChoose(getSourceId(), player.getId(), game)) {
+                            if (player.chooseTarget(Outcome.Benefit, target, ability, game)) {
+                                Permanent targetPermanent = game.getPermanent(target.getFirstTarget());
+                                boolean mutateUnder = player.chooseUse(Outcome.Neutral,
```
"mutate under" or "mutate over" are not "modes" nor are they choices you make as you cast the spell. You choose over or under as the spell resolves.

ninthworld replied:
This is something I was going over when thinking about copying a mutate spell (like with Volo), and I thought I read somewhere that every aspect of the spell needs to be the same, which I thought included the over/under. However, I can't find anything saying that in the rules, so I'm pretty sure you're right.

awjackson replied:
> 702.140c As a mutating creature spell resolves, if its target is legal, it doesn’t enter the battlefield. Rather, it merges with the target creature and becomes one object represented by more than one card or token (see rule 723, “Merging with Permanents”). The spell’s controller chooses whether the spell is put on top of the creature or on the bottom. The resulting permanent is a mutated permanent.

theelk801 replied:
How does it currently work for a card like [[Momentary Blink]] or a creature with persist? Any time a mutated permanent leaves and re-enters it needs to come back as each individual creature separately.

ninthworld replied:
> How does it currently work for a card like [[Momentary Blink]] or a creature with persist? Any time a mutated permanent leaves and re-enters it needs to come back as each individual creature separately.

Momentary Blink should be covered by my `setupTestMutateToExileAndBack` unit tests, however, Persist and Regenerate seem to need more work to implement properly.

ninthworld replied:
> Momentary Blink should be covered by my `setupTestMutateToExileAndBack` unit tests, however, Persist and Regenerate seem to need more work to implement properly.

Persist, Undying, and Regenerate are all working now

awjackson commented:
I think `MutateAbility` needs to be implemented as a subclass of `SpellAbility` (compare to bestow) rather than as a subclass of `AlternativeSourceCosts`. The main reason is that casting a creature spell for its mutate cost changes the spell into a targetted spell, which has pretty widespread effects both from a game rules and a UI standpoint. It's better to leverage the core xmage functionality for targetting (e.g. checking whether a valid target exists before you are allowed to attempt to cast the spell) than try to reimplement it from scratch like you are doing here.

ninthworld commented:
> I think `MutateAbility` needs to be implemented as a subclass of `SpellAbility` (compare to bestow) rather than as a subclass of `AlternativeSourceCosts`. The main reason is that casting a creature spell for its mutate cost changes the spell into a targetted spell, which has pretty widespread effects both from a Magic rules and xmage UI standpoint. It's better to leverage the core xmage functionality for targetting (e.g. checking whether a valid target exists before xmage highlights the card as usable and allows you to attempt to cast it) than try to reimplement it from scratch like you are doing here.

Great suggestion! This simplified the logic of the `MutateAbility` significantly and allowed me to get Brokkos casting from the graveyard working (albeit with a very improper implementation that will still need to be fixed)

teskogi commented:
Wow! Nice to see someone else taking a crack at mutate. I tried my hand at it a while ago. I got it working against a couple of test cases I made (sans blinking the mutated permanent), but I don't think the UI quite accepted what I was trying to do. Anyway, I'm attaching the patch files for it, hopefully they'll be of some use.
[mutatePatches.zip](https://github.com/magefree/mage/files/8383791/mutatePatches.zip)

teskogi commented:
Hi ninthworld, I've had some issues getting your test cases to work on my local copy. Some of it seems to be some staleness of your branch, which the below patch should take care of. 
[FixMutate.zip](https://github.com/magefree/mage/files/8536420/FixMutate.zip)

But some of the tests fail sometimes and not others. Perhaps there are some race conditions at play somewhere.

ninthworld commented:
> Hi ninthworld, I've had some issues getting your test cases to work on my local copy. Some of it seems to be some staleness of your branch, which the below patch should take care of. [FixMutate.zip](https://github.com/magefree/mage/files/8536420/FixMutate.zip)
> 
> But some of the tests fail sometimes and not others. Perhaps there are some race conditions at play somewhere.

I believe the reason tests are failing sometimes (rarely) is the computer randomly using all of a single color of land before the card that needs that color has been cast.

sprangg commented:
Is there any reason to not merge this into beta so that we can start playing around with it and seeing how it goes? And reporting any issues encountered, of course.

Great work btw, one of xmage's most requested features if not the most.

theelk801 commented:
I haven't had a chance to look over this code yet but I'm definitely down to try and make it work. I've already spent some time refactoring exile effects and there are other things that need changing ("exile target permanent until this leaves the battlefield", meld, etc) so I think it's worth the effort if we organize it well. I'll probably be able to put even more focus into it after CLB is done.

teskogi commented:
I got this to pass checks again. PersistAbility and UndyingAbility needed to be patched. I also changed MutateCommanderTest to fix the mana issues. I added Ignore to a few of the MutateTest tests, but they do need to be completed.

[8733fixed.zip](https://github.com/magefree/mage/files/8792605/8733fixed.zip)

ninthworld commented:
> I got this to pass checks again. PersistAbility and UndyingAbility needed to be patched. I also changed MutateCommanderTest to fix the mana issues. I added Ignore to a few of the MutateTest tests, but they do need to be completed. I also fixed some issues with GameState due to the implementation of Aeon Engine.
> 
> [8733fixed.zip](https://github.com/magefree/mage/files/8792625/8733fixed.zip)

Thanks! I haven't worked on this PR in a bit since I've had some other stuff going on. Hopefully once I get some time I'll be able to come back to it with some fresh eyes.

teskogi commented:
Some of the cards in the IkoriaLairOfBehemoths set also need NON_FULL_USE_VARIOUS to pass verification. I've updated the patch.
[8733fixed.zip](https://github.com/magefree/mage/files/8793907/8733fixed.zip)

ninthworld commented:
After some time away, I decided to take another swing at implementing mutate. Then I ran into all the same issues that I had implementing this version and remembered why I had to make all the compromises I did...

So instead, I cleaned up this one, got the rest of the tests working, squashed a bunch of lingering bugs, and finally got it to pass all build checks. I'm sure there are abilities/effects that still don't work with mutate, but I don't know what they are and I think its going to take some community testing to find them.

sprangg commented:
Thank you for your hard work! Mutate for beta testing let's goooo?

Alex-Vasile commented:
> Thank you for your hard work! Mutate for beta testing let's goooo?

As much as it pains me to say it. This one really should wait until after the next release (🤞) to be merged. Doing it now would only further push the release back.

jeffwadsworth commented:
> > Thank you for your hard work! Mutate for beta testing let's goooo?
> 
> As much as it pains me to say it. This one really should wait until after the next release (🤞) to be merged. Doing it now would only further push the release back.

There are no new releases.  The Beta is now the “release”.  As the coder noted, the community can help find bugs.

jeffwadsworth commented:
Great work on this ninthworld.  Hopefully it works out.

Alex-Vasile reviewed (approved):
Looks good as far as I can tell. Will need to actually play it to find whatever bugs are hiding

ninthworld commented on Mage/src/main/java/mage/abilities/keyword/MutateAbility.java:
```diff
@@ -1,39 +1,191 @@
+    }
+
+    @Override
+    public boolean askToActivateAlternativeCosts(Ability ability, Game game) {
```
I assume you meant to tag `mutatePermanent()` in `GameImpl` class, I flattened it a bit more.

ninthworld commented on Mage/src/main/java/mage/game/GameImpl.java:
```diff
@@ -3763,4 +3763,123 @@ public String toString() {
+                                if (target.equals(permanentId)) {
+                                    targets.replaceMutatedTarget(permanentId, newPermanent.getId());
+                                    break;
+                                }
```
So as its written, it should be able to handle targeting the same creature multiple times, but I added some unit tests to address it and they are failing due to some lingering `zoneChangeCounter` issues. 

Those issues are lingering because zoneChangeCounters are basically like black magic to me and I don't understand why they are stored everywhere, how to know what they are/should be, and why it is that when I manipulate the zoneChangeCounters to fix these two new unit tests, it breaks 4 others...

teskogi commented:
Any ideas why the tests fail now? I removed the calls to assertAllCommandsUsed(), but that didn't seem to solve the issue.

awjackson commented:
The six conflicting files probably have something to do with the build failing...

I finally took a close look at this and I really don't like the approach of replacing a permanent that's already on the battlefield with a new permanent with a different UUID when you "mutate over". You have added workarounds in Target and TargetPointer, but between the core rules engine, the GUI, common abilities/effects, and card-specific code there are probably thousands and thousands of places where game objects are expected to have unchanging UUIDs and things will break if they don't. Attachments (Auras, equipment) are just the first one I came up with off the top of my head.

I think we need a new approach where the first creature on the battlefield is always the base permanent, and when you "mutate over", rather than replacing the base permanent with a new permanent, you just make the existing permanent get its characteristics from the new card, similar to how a TDFC transforming or a permanent becoming a copy of something works.

ninthworld commented:
> The six conflicting files probably have something to do with the build failing...
> 
> I finally took a close look at this and I really don't like the approach of replacing a permanent that's already on the battlefield with a new permanent with a different UUID when you "mutate over". You have added workarounds in Target and TargetPointer, but between the core rules engine, the GUI, common abilities/effects, and card-specific code there are probably thousands and thousands of places where game objects are expected to have unchanging UUIDs and things will break if they don't. Attachments (Auras, equipment) are just the first one I came up with off the top of my head.
> 
> I think we need a new approach where the first creature on the battlefield is always the base permanent, and when you "mutate over", rather than replacing the base permanent with a new permanent, you just make the existing permanent get its characteristics from the new card, similar to how a TDFC transforming or a permanent becoming a copy of something works.

I completely agree with you, however, due to the way permanents are implemented in the core engine, it is not possible as far I have tried. For example, if you have a token permanent on the battlefield, and you mutate over it with a regular permanent, the object on the battlefield would still be a `TokenPermanent` with your suggestion. There is no way to make it a `CardPermanent` without just making a new object.

And the reason the tests are failing is because some logic dealing with the `ZoneChangeCounters` is incorrect and causing sync issues between `Card` `MageObjectReference` and the zoneChangeCounters in `GameState` and through multiple tries I can't seem to understand how to get it to work correctly.

awjackson commented:
> I completely agree with you, however, due to the way permanents are implemented in the core engine, it is not possible as far I have tried. For example, if you have a token permanent on the battlefield, and you mutate over it with a regular permanent, the object on the battlefield would still be a `TokenPermanent` with your suggestion. There is no way to make it a `CardPermanent` without just making a new object.

ExpensiveKoala commented:
What is still pending on this PR? What is holding it back from merging (other than merge conflicts)

thefnitalian commented:
What is the status on this?

ExpensiveKoala commented:
This PR's implementation has some issues with copies of on-top mutated creatures. To replicate:
Have a [[Scute Swarm]] and 5+ lands under your control. Mutate a [[Migratory Greathorn]] on TOP of the Scute Swarm. The Migratory Greathorn's mutate triggers and you get a land, which triggers Scute Swarm to create a copy of itself since you'd have 6+ lands. However, any subsequent lands you play will only trigger the landfall of the copies and not the original anymore.
This doesn't seem to happen when you mutate the Migratory Greathorn on bottom of the Scute Swarm.

xenohedron commented:
@ninthworld thanks for your work on this, especially the comprehensive tests. Studying this implementation really helped me eventually develop a more sustainable implementation in #14900
